api/api.go

```diff
@@ -55,24 +63,71 @@ func registerValidators() error {
 	return nil
 }
 
-func registerRoutes(ws *gin.Engine, versionsRegistry data.VersionsRegistryHandler) error {
+func registerRoutes(ws *gin.Engine, versionsRegistry data.VersionsRegistryHandler, credentialsConfig config.CredentialsConfig) error {
 	versionsMap, err := versionsRegistry.GetAllVersions()
 	if err != nil {
 		return err
 	}
 
 	for version, versionData := range versionsMap {
 		versionGroup := ws.Group(version)
-
 		for path, group := range versionData.ApiHandler.GetAllGroups() {
 			subGroup := versionGroup.Group(path)
-			group.RegisterRoutes(subGroup)
+			group.RegisterRoutes(subGroup, versionData.ApiConfig, getAuthenticationFunc(credentialsConfig))
 		}
 	}
 
 	return nil
 }
 
+func getAuthenticationFunc(credentialsConfig config.CredentialsConfig) gin.HandlerFunc {
+	var hasher hashing.Hasher
+	var err error
+	hasher, err = factory.NewHasher(credentialsConfig.Hasher.Type)
+	if err != nil {
+		log.Warn("cannot create hasher from config. Will use Sha256 as default", "error", err)
+		hasher = sha256.Sha256{} // fallback in case the hasher creation failed
+	}
+
+	accounts := gin.Accounts{}
+	for _, pair := range credentialsConfig.Credentials {
+		accounts[pair.Username] = pair.Password
+	}
+
+	authenticationFunction := func(c *gin.Context) {
+		user, pass, ok := c.Request.BasicAuth()
+		if !ok {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, data.GenericAPIResponse{
+				Data:  nil,
+				Error: "This endpoint requires Basic Authentication",
+				Code:  data.ReturnCodeRequestError,
+			})
+			return
+		}
+
+		userPassword, ok := accounts[user]
+		if !ok {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, data.GenericAPIResponse{
+				Data:  nil,
+				Error: "Username does not exist in configuration",
```

Comment:
`Username does not exist` should be enough. We should avoid giving too much info to whoever is trying this url. Plus, we may want in the future to have a more advanced authentication form than using files

Reply:
replaced